### ISSUE TITLE
[Techstack-beta] Not being built as a Nix build output

### ISSUE THREAD

**Issue by samueldr** [NixOS/nixos-homepage#1296]
Not being able to handle the build of the website with a correct Nix build output gives mixed messages about Nix and its ecosystem, imo.

I was surprised earlier when testing the site and trying to check something that `nix-build` (or `nix build`) won't build the site anymore.

* * *

Refs:

Previous implementation (calling `nix build`):

 - https://github.com/NixOS/nixos-homepage/blob/91c8dce4ca60e875b0302c0e47c0d5ba4f8a55ad/.github/workflows/master.yml#L30-L35

New implementation (running non-Nix package management outside of a Nix build):

 - https://github.com/NixOS/nixos-homepage/blob/90f8e1f155b04221d49596e84fa69b418b610eb9/.github/workflows/ci.yml#L44-L50

**Comment by garbas** [comment]
Yes, this was decided for now. It might even stay this way. We'll evaluate this after the migration is complete and after some time of using the website.

**Comment by thilobillerbeck** [comment]
Currently this is hindered by one component, namely `Astro-Icon` which fetches Icons on the fly during build time to save up on disk space (you don't want the whole Iconify catalogue on your machine). I want to swap this out for the iconify-tailwind plugin anyway (so Iconsets are a normal npm dependency and thus can integrate with `build{Node/NPM/Yarn}*` so this is not a big deal and we for sure can realize this.

The huge problem we **could face** in the future will be foreign content which would be fetched from other content sources like APIs, thumbnails for things, discourse etc. This is already a problem that struck me on the NixCon website, where the thumbnails for the recordings are dynamically fetched and compressed to save on loading time.
You can integrate everything into a flake and fetch it beforehand for sure, but as those kinds of features get more complex. And this is not an Astro specific thing either, this is just how modern static site generators tend to work.

Since the Nix landscape is big and I think we are not the only ones with this problem, I would highly encourage everyone stumbling upon this to suggest a solution that tackles the problem mentioned in the paragraph before (without disabling the sandbox :P). I think solving this issue would be a great win for the community and would solve lots of similar problems. 

Btw. on a **personal note**: I don't think this is sending mixed messages. Imo. we should encourage people to integrate nix as they want it to be in their projects. This is the beauty of nix after all, you can have either a shell, or tests or even a build, but you have the choice of what fits in your project and what does not. 

Please see the above as an explaination and not as a proper plan we have. This is just intended to explain problems we could face in the future.

I am really happy for any suggestions and a healthy discussion in this thread.

**Comment by thilobillerbeck** [comment]
Currently there is one dependency which requires network connectivity to fetch icons at build time, but this will be replaced shortly. Then a nix build would be possible.

Unfortunately if we want to fetch resources like APIs etc. in the future, we would need network access during build time which would either mean disableing the sandbox during build time or not building through nix. If there is any solution I don't have in mind, just suggest it here. While downloading thing upfront through a flake would be possible, it circumvents many features modern SSGs have like image fetching and optimization etc.

**Comment by thilobillerbeck** [comment]
Btw. since issues with sharp etc. have been resolved this can now be built with nix, though since flakes don't support attributes it is not pure in production. If I can find a good solution to circumvent this problem, it also will be pure there.
